components/core/src/ffi/search/WildcardToken.cpp

```diff
@@ -0,0 +1,209 @@
+                return false;
+            }
+        }
+        return true;
+    }
+
+    /**
+     * To check if the token could be static text, formally, we need to check if
+     * the token matches the complement of all variable schemas ORed together
+     * (~((schema1)|(schema2)|...). Another way of looking at this is if the
+     * token contains anything which indicates it's definitely a variable, then
+     * it can't be static text.
+     */
+    static bool could_be_static_text (string_view query, size_t begin_pos, size_t end_pos) {
+        bool is_escaped = false;
+        bool contains_alphabet = false;
+        for (size_t i = begin_pos; i < end_pos; ++i) {
+            auto c = query[i];
+            if (is_escaped) {
+                is_escaped = false;
+            } else if ('\\' == c) {
+                is_escaped = true;
+            } else if (is_decimal_digit(c)) {
+                return false;
+            } else if (is_alphabet(c)) {
+                contains_alphabet = true;
+            }
+        }
+
+        if (begin_pos > 0 && '=' == query[begin_pos - 1]) {
+            if ('?' == query[begin_pos] && contains_alphabet) {
+                // "=?...<alphabet>..." must be a variable since
+                // 1. '?' would only be included in the variable token if it
+                //    was treated as a non-delimiter, and
+                // 2. an '=' followed by non-delimiters and an alphabet is
+                //    definitely a variable.
+                return false;
+            }
+        }
+
+        return true;
+    }
+
+    template <typename encoded_variable_t>
+    WildcardToken<encoded_variable_t>::WildcardToken (
+            string_view query, size_t begin_pos, size_t end_pos
+    ) : QueryToken(query, begin_pos, end_pos), m_has_prefix_star_wildcard('*' == query[begin_pos]),
+        m_has_suffix_star_wildcard('*' == query[end_pos - 1])
+    {
+        auto token = string_view(query.cbegin() + begin_pos, end_pos - begin_pos);
+        if (could_be_int_var<encoded_variable_t>(token)) {
+            m_possible_variable_types.push_back(TokenType::IntegerVariable);
+        }
+        if (could_be_float_var<encoded_variable_t>(token)) {
+            m_possible_variable_types.push_back(TokenType::FloatVariable);
+        }
+        if (could_be_static_text(query, begin_pos, end_pos)) {
+            m_possible_variable_types.push_back(TokenType::StaticText);
+        }
+        // Value must contain a wildcard and a non-delimiter, so it can be a
+        // dictionary variable
+        m_possible_variable_types.push_back(TokenType::DictionaryVariable);
+
+        m_current_interpretation_idx = 0;
+    }
+
+    template <typename encoded_variable_t>
+    bool WildcardToken<encoded_variable_t>::add_to_logtype_query (string& logtype_query) const {
+        auto current_interpretation = m_possible_variable_types[m_current_interpretation_idx];
+        if (TokenType::StaticText == current_interpretation) {
```

Comment:
Quick question: why we don't need to consider the case for begin = wildcard?

Reply:
Added more comments to clarify.